ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/GenesisGenerator.java

```diff
@@ -0,0 +1,148 @@
+import static tech.pegasys.artemis.util.config.Constants.SECONDS_PER_DAY;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Predicate;
+import org.apache.logging.log4j.Level;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.artemis.datastructures.blocks.BeaconBlockBody;
+import tech.pegasys.artemis.datastructures.blocks.BeaconBlockHeader;
+import tech.pegasys.artemis.datastructures.blocks.Eth1Data;
+import tech.pegasys.artemis.datastructures.operations.Deposit;
+import tech.pegasys.artemis.datastructures.operations.DepositData;
+import tech.pegasys.artemis.datastructures.state.BeaconState;
+import tech.pegasys.artemis.datastructures.state.BeaconStateWithCache;
+import tech.pegasys.artemis.datastructures.state.Validator;
+import tech.pegasys.artemis.util.SSZTypes.SSZList;
+import tech.pegasys.artemis.util.bls.BLSPublicKey;
+import tech.pegasys.artemis.util.hashtree.HashTreeUtil;
+
+public class GenesisGenerator {
+  private final MerkleTree<Bytes32> depositMerkleTree =
+      new MerkleTree<>(DEPOSIT_CONTRACT_TREE_DEPTH);
+  private final BeaconState state = new BeaconState();
+  private final Map<BLSPublicKey, Integer> keyCache = new HashMap<>();
+  private final long depositListLength = ((long) 1) << DEPOSIT_CONTRACT_TREE_DEPTH;
+  private final SSZList<DepositData> depositDataList =
+      new SSZList<>(DepositData.class, depositListLength);
+
+  public GenesisGenerator() {
+    BeaconBlockHeader beaconBlockHeader = new BeaconBlockHeader();
+    Bytes32 latestBlockRoot = new BeaconBlockBody().hash_tree_root();
+    beaconBlockHeader.setBody_root(latestBlockRoot);
+    state.setLatest_block_header(beaconBlockHeader);
+  }
+
+  public void addDepositsFromBlock(
+      Bytes32 eth1BlockHash, UnsignedLong eth1Timestamp, List<? extends Deposit> deposits) {
+    updateGenesisTime(eth1Timestamp);
+
+    final Eth1Data eth1Data = state.getEth1_data();
+    eth1Data.setBlock_hash(eth1BlockHash);
+    eth1Data.setDeposit_count(UnsignedLong.valueOf(depositDataList.size() + deposits.size()));
+
+    // Process deposits
+    deposits.forEach(
+        deposit -> {
+          if (BeaconStateUtil.DEPOSIT_PROOFS_ENABLED) {
+            calculateDepositProof(deposit);
+          }
+          STDOUT.log(Level.DEBUG, "About to process deposit: " + depositDataList.size());
+          depositDataList.add(deposit.getData());
+
+          // Skip verifing the merkle proof as we'll only generate one at the end
+          // We do still verify the signature
+          process_deposit_without_checking_merkle_proof(state, deposit, keyCache);
+
+          processActivation(deposit);
+        });
+  }
+
+  private void processActivation(final Deposit deposit) {
+    final Integer index = keyCache.get(deposit.getData().getPubkey());
+    if (index != null) { // Could be null if the deposit was invalid
```

Comment:
(nit) we can avoid nesting:
```suggestion
    if (index == null) return  // Could be null if the deposit was invalid
```

Reply:
Good point.